nearcore/src/config.rs

```diff
@@ -1137,65 +1137,73 @@ pub fn get_config_url(chain_id: &str) -> String {
 
 #[derive(thiserror::Error, Debug)]
 pub enum FileDownloadError {
-    #[error("Failed to download the file: {0}")]
-    HttpError(#[from] hyper::Error),
-    #[error("Failed to open file: {0}")]
+    #[error("{0}")]
```

Comment:
This should be `#[error(transparent)]` I think.

---

Pre-existing, but in general `thiserror` messages should _not_ format in the cause error's description into its own. Instead `#[source]` mechanism should be utilized to specify the causation chain. This is how the ecosystem expects the errors to be, making it least likely to get duplicated output like you had described in the PR description. I have, in the past, also written [a block post on error handling](https://kazlauskas.me/entries/errors). While this is not directly applicable to this discussion, it kind of demonstrates what a conventional thiserror use looks like.

Reply:
Ah yeah, I actually tried this originally but it gives the same repition of info. Looks like something wrong with hyper itself?

```rust
println!("orig {}", &e);
let mut source = e.source();

while let Some(err) = source {
  println!("source {}", err);
  source = err.source();
}
```

this prints out:

```
orig error trying to connect: tcp connect error: Connection refused (os error 111)
source tcp connect error: Connection refused (os error 111)
source Connection refused (os error 111)
```

So just using "{0}" is the best workaround for that I can think of here. But as for the others, using #[source] instead of formatting it explicitly seems way better yeah, I like the blog post! updated the other ones to use #[source] now, and checked that the output is still good. (Last one is a little tricky since there are really two sources, but we can just pick the first one I guess)

Reply:
I see. I think it might be a good idea to report this to hyper.

Reply:
https://github.com/hyperium/hyper/issues/2732